### ISSUE TITLE
Android release preparations

### ISSUE THREAD

**Issue by ihhub** [ihhub/fheroes2#6008]
### Preliminary checks

- [X] I've checked that there aren't [**other open issues**](https://github.com/ihhub/fheroes2/issues?q=is%3Aissue) on the same topic.

### Describe the problem requiring a solution

Even having a very basic Android build we are far away from a proper release for this platform. We need to make multiple changes and adjustments in order to make it user-friendly and ready to be published in App Store.

### Describe the possible solution

- [x] lock the application in landscape mode. Right now it is possible to rotate the view.
- [x] add compilation flags which are present in other builds, especially for warnings.
- [x] make `files/data` directory and compiled translations to be a part of the package to avoid extra resource copying even by developers.
- [x] Android filesystem is case-sensitive. Right now the code does not handle it properly. The logic present for case-sensitive filesystems is not applicable for Android due to a fact that an Android application has access (even for reading) only to certain directories.
- [x] add proper support of internal and external storages.
- [x] evaluate the need of permissions needed for the application.
- [x] add Java code to verify the presence of files and possibly download a demo version and unzip the downloaded archive into the specific location.
- [x] add a script to setup Android build for developer. Include needed paths to `.gitignore` file (looks like it is done).
- [x] fix MIDI music for Android. Android has native support of it.
- [x] fix resolution detection. In the Android devices the height is bigger than width.
- [x] add package signing as a part of CI/CD pipeline.
- [x] finalize minimum Android version requirements for the application.
- [x] right mouse click simulation
- [x] add a UI element to toggle between full-screen and windowed modes
- [x] connect system audio level to game's level of music or audio or both
- [x] add code to open keyboard for typing words like it is done for Switch and Vita

**Comment by oleg-derevenetz** [comment]
> Android filesystem is case-sensitive. Right now the code does not handle it properly. The logic present for case-sensitive filesystems is not applicable for Android due to a fact that an Android application has access (even for reading) only to certain directories.

If we are going to distribute APK via direct download, we can just use the [MANAGE_EXTERNAL_STORAGE](https://developer.android.com/training/data-storage/manage-all-files) permission and screw the Google "best practices".

> Android has native support of it.

I'm not sure about that. `SDL_Mixer` [has support](https://github.com/libsdl-org/SDL_mixer/tree/main/src/codecs/native_midi) for native MIDI on MacOS X and Windows, but not on Android. It can be built with TiMidity though (just like I did for our Android package), but in this case it requires additional files (GUS patches).

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I see no issues not to follow "best practices" but make the application more user-friendly in general.

**Comment by oleg-derevenetz** [comment]
> I see no issues not to follow "best practices" but make the application more user-friendly in general.

The problem is that in this case app becomes less suitable for Google Play :) So we have to choose between Google Play and ease of use in this case. Also people are often scared by permission requests.

**Comment by ihhub** [comment]
We should enable software cursor rendering but we need to fix #5778 in order to make it work.

**Comment by oleg-derevenetz** [comment]
@ihhub

> We should enable software cursor rendering

Why? What's the purpose of the mouse cursor on the touchscreen?

**Comment by ihhub** [comment]
@oleg-derevenetz , we need to choose one of the way to control the game on Android devices:
- full touchscreen. In this case we need to add extra touch actions as now it is easy to miss a button or a cell in a battle
- emulation of cursor and clicks in the game. Showing a visible cursor to have high precision clicks.

I was thinking to try the second approach but looks like we have no choice but to stick to the first one. We can add some touch actions to avoid difficulties of controlling the game.

**Comment by ihhub** [comment]
My artistic skills are not up to date but:
![image](https://user-images.githubusercontent.com/19829520/197531356-fbe1d751-e2e4-4a78-8f70-96b624c36ec4.png)

**Comment by ihhub** [comment]
I've seen such games being played by other people. Of course it doesn't need to be black background and etc. but many games use this concept. If we stick to it we don't need to change much code in the game regarding controls.

**Comment by oleg-derevenetz** [comment]
@ihhub Well, looks promising. Does SDL have some built-in support for this mode?

**Comment by ihhub** [comment]
@oleg-derevenetz , I do not know. We need to investigate this🙂

**Comment by oleg-derevenetz** [comment]
@ihhub

> left click simulation

I suppose you mean RIGHT click simulation :) Left click is simulated just fine using the single touch.

**Comment by ihhub** [comment]
@oleg-derevenetz , yes! I need more sleep :)

**Comment by poVoq** [comment]
Any idea why copying the map files fails? I got the game up and running, but android refuses to copy the map files and shows some strange java error and the .MP2 files end up as 0 byte .MP2.mpeg files in the folder I need to copy them to.

**Comment by oleg-derevenetz** [comment]
Hi @poVoq how exactly are you trying to copy them? I usually use either Total Commander on the Android device itself or Windows Explorer when the device is attached to Windows machine using USB cable. It seems that the program you are using perceives the map files as multimedia files because of their MP2 extension. You can try to pack them into the ZIP archive, copy this ZIP archive to the device and then extract these files to the `MAPS` directory using Total Commander.

**Comment by poVoq** [comment]
I tried using the SimpleFiles app from f-droid. Other Apps do not seem to allow accessing the /android/data/ folder at all. I will try over a usb connection, but I think that system folder is not easily accessible that way either?

**Comment by oleg-derevenetz** [comment]
@poVoq this is `Android/data/..` on the shared storage, i.e. full path is `/sdcard/Android/data/...`. Any file manager, including the Android preinstalled "Files" app is able to access it and write to it.

**Comment by poVoq** [comment]
In the standard files app from LinageOS it is totally inaccessible and the SimpleFiles app needs to ask special permission to access it. Copying via usb seems to work, but it is super slow and breaks down every new few minutes. Normally that does not happen when copying files to normal directories like the /downloads folder.

**Comment by zenseii** [comment]
> Copying via usb seems to work, but it is super slow and breaks down every new few minutes. Normally that does not happen when copying files to normal directories like the /downloads folder.

This probably happens because the device has been inactive for a moment and therefore goes into some rest mode. I transferred each directory separately.

**Comment by oleg-derevenetz** [comment]
@poVoq It's strange that interaction with this folder is broken on the LineageOS , because the main purpose of this folder is to allow user to exchange files with the installed apps.

**Comment by poVoq** [comment]
My impression is that this folder is meant for storing app specific configuration files etc. and is not mean to be accessed manually at all.

**Comment by oleg-derevenetz** [comment]
@poVoq Nope, there is separate folder for this - `/data/data/org.fheroes2` for this, so-called "private app folder". This folder is not accessible from outside (if phone is not in debug mode).

**Comment by Laserlicht** [comment]
@poVoq Maybe this problem? https://www.reddit.com/r/Android/comments/j3zgmm/managing_files_in_the_androiddata_folder_on/

@oleg-derevenetz Is it possible to run the app in [immersive mode](https://developer.android.com/develop/ui/views/layout/immersive)?

**Comment by oleg-derevenetz** [comment]
@Laserlicht

> Is it possible to run the app in [immersive mode](https://developer.android.com/develop/ui/views/layout/immersive)?

Yes, of course. Just replace `fullscreen = off` by `fullscreen = on` in `fheroes2.cfg`.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I am thinking we need to make full screen for Android devices by default. However, having a special option for this is in my to-do list.

I also found a crash: use drag and drop to split a stack and when you press okay, the application crashes.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub

> I also found a crash: use drag and drop to split a stack and when you press okay, the application crashes.

This crash is not related to Android, it is due to an assertion failure:

```
Assertion failed: (0), function addNewTroopsToFreeSlots, file army.cpp, line 813.
```

**Comment by oleg-derevenetz** [comment]
> I am thinking we need to make full screen for Android devices by default

As for me personally, I'd like to have a choice, because I'm used to seeing the clock in the top bar to keep track of time. But this is a matter of taste.

**Comment by oleg-derevenetz** [comment]
> https://www.reddit.com/r/Android/comments/j3zgmm/managing_files_in_the_androiddata_folder_on/

OMG, WTF. Another shitty design decision from Google under the sauce of "security". It looks like that they want to have only messengers, Facebook and Instagram on the platform. Apps that work with files? Meh.

**Comment by ihhub** [comment]
Let's stick for Android 10-12 for now. All fancy Android 13 devices won't come up so quickly.

**Comment by ihhub** [comment]
I suggest for **right** click (I am almost falling asleep again) we can use multi-touch event handled by SDL2. This is either [SDL_DollarGestureEvent](https://wiki.libsdl.org/SDL_DollarGestureEvent) or [SDL_MultiGestureEvent](https://wiki.libsdl.org/SDL_MultiGestureEvent).

This is an easy workaround to make some sort of workable touchscreen version of the game. What do you think?

**Comment by ihhub** [comment]
Actually we can modify this method `void LocalEvent::HandleTouchEvent( const SDL_TouchFingerEvent & event )` to handle multi-finder gestures.

**Comment by zenseii** [comment]
A quick solution to perform a right click is to hold one finger, the thumb, on the object of interest and the press another finger, the other thumb, anywhere else on the screen. When the second finger is released the right click stops.

**Comment by oleg-derevenetz** [comment]
> Let's stick for Android 10-12 for now. All fancy Android 13 devices won't come up so quickly.

In principle, this could be solved by an additional "toolset" Activity with some GUI that will be able to, say, open ZIP file with OG data files and maps from the SD card using the system file picker with `ACTION_OPEN_DOCUMENT`, and extract this ZIP file to the `ExternalFilesDir`. It may look like additional icon on the Launchpad. This additional Activity can also be used to download the demo version assets if user doesn't have a purchased copy of the original game. Maybe someone with experience in developing GUI applications for Android could do it.

**Comment by oleg-derevenetz** [comment]
> connect system audio level to game's level of music or audio or both

I'm not sure that we need to do something in this regard. Currently the system audio volume is a "master volume" and can be adjusted in the usual way by the volume buttons on the side face of the device. Music and effects volume are adjusted in the game itself relative to this master volume. I believe that everything already works as it should.

**Comment by ihhub** [comment]
> > connect system audio level to game's level of music or audio or both
> 
> I'm not sure that we need to do something in this regard. Currently the system audio volume is a "master volume" and can be adjusted in the usual way by the volume buttons on the side face of the device. Music and effects volume are adjusted in the game itself relative to this master volume. I believe that everything already works as it should.

It's a fair point. I will mark it as completed.

**Comment by ihhub** [comment]
Regarding minimum Android version: I saw that players were capable to play the game on Android 8 but it doesn't work on Android 7. At least we should highlight recommended Android version to avoid any complains from users.

**Comment by oleg-derevenetz** [comment]
@ihhub I have the Android 6 tablet and was able to run & play the fheroes2 just fine (that was some time ago, I'll try this again tomorrow, but I suppose it will work). In fact, the minimum supported Android API version now is 19, which means Android 4.4 KitKat  - and that's only because this is the [minimum version supported by the latest NDK](https://github.com/android/ndk/wiki/Changelog-r24), the SDL itself should work on SDK 16 (Android 4.1).

**Comment by ihhub** [comment]
Wow, @Branikolog wasn't able to run fheroes2 on his device which is Android 7 based.

**Comment by oleg-derevenetz** [comment]
@ihhub Yes, I just tried it and it won't work on Android 6. The reason is Apache commons-io library, which modern version uses Java language features that Android 6 didn't support yet, for example, lambdas. If I use an older [version 2.8.0](https://mvnrepository.com/artifact/commons-io/commons-io), fheroes2 runs on Android 6 just fine. I can switch to commons-io 2.8.0 and set Android 6 as the minimum supported version. What do you think?

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , lets downgrade it since we don't have much code in Java.

**Comment by oleg-derevenetz** [comment]
Hi @ihhub I implemented the proper desugaring of advanced Java language features instead.

BTW please note that only 64-bit platforms are currently supported (arm64-v8a and x86_64). 32-bit platforms would just uselessly take up space, because all 32-bit devices should be at the landfill by now :)

**Comment by oleg-derevenetz** [comment]
Tablet experience is very nice BTW. On 6'' phone everything is too small, even if 640x480 is selected.

**Comment by ihhub** [comment]
> Tablet experience is very nice BTW. On 6'' phone everything is too small, even if 640x480 is selected.

Now imagine people trying to set 2000x1000 resolution on their phones....

**Comment by ihhub** [comment]
Since we locked the screen position into landscape mode resolution detection works fine.

**Comment by oleg-derevenetz** [comment]
Hi @zenseii

> We have a user who's willing to test on Android 5.1: https://heroes2.forumactif.com/t1013-android-build-quick-install-guide-for-noobs#9203

There are most likely two issues:

1. Currently Android APK has the "hard limit" for the lowest Android version, which is currently 6.0. So this user will not be able to install `fheroes2.apk` on 5.x anyway.
2. His `INSTALL_FAILED_NO_MATCHING_ABIS` error may be because he has 32-bit Android, while `fheroes2.apk` is currently built only for 64-bit architectures (`arm64-v8a` and `x86_64`).

I can later create a separate PR with support for the lower minimal Android version (5.1) and 32-bit architectures for ancient devices.

**Comment by oleg-derevenetz** [comment]
@zenseii

> I looked up what I believe is the CPU from reference sites and it says 64-bit on Intel's own page (just scroll to the bottom of the page

Yeah, but even if CPU itself supports 64-bit, it doesn't mean that vendor runs 64-bit Android on it. It could as well run 32-bit Android.

**Comment by oleg-derevenetz** [comment]
@zenseii he can try the APK from #6324. Please note that this is the non-optimized debug build, it is only suitable to check the ability to basically run on his device, and not to evaluate the performance and speed.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz and @zenseii , should we include scripts for Linux and Windows to download a demo version of the game for the Android archive? I have a feeling people would ask us about the way to download the demo.

**Comment by oleg-derevenetz** [comment]
@ihhub

> should we include scripts for Linux and Windows to download a demo version of the game for the Android archive?

We can simply specify the URL of the demo version in the README. Making a special script just to download and unpack one ZIP file is, in my opinion, excessive. It was convenient for installation maintenance in these systems, because it was able to copy demo files to the appropriate places, but on Android people should copy files themselves.

**Comment by ihhub** [comment]
I am okay to at least modify the README file to include the link so users at least know where to get it. However, we should specify that we have no affiliation with the host of this demo version.

**Comment by Laserlicht** [comment]
It would be great if we could get the app in the app store soon. Since the manual update on my device is not really great (app can not be overwritten, must be deleted and reinstalled with the additional files).

Maybe first at [F-Droid](https://f-droid.org/en/contribute/). This is usually used by more experienced users. It is not so bad if it is not yet so perfect.

**Comment by oleg-derevenetz** [comment]
@Laserlicht Debug builds from PR artifacts and release builds from releases section (including the "latest snapshot") are signed by different signatures. If you don't need to install builds from PRs, then just avoid this :)

**Comment by Laserlicht** [comment]
Hi @zenseii & @oleg-derevenetz Thank you very much. 😊

**Comment by oleg-derevenetz** [comment]
@Laserlicht I doubt that Google Play will accept this app in its current form, at least we should add some sort of "launcher" activity with minimal user-friendly UI that should help the unprepared user to copy OG files in place.

**Comment by Laserlicht** [comment]
@oleg-derevenetz 

Wouldn't it be even better (besides the possibility to extract the files from a zip) if we integrate the [innoextract](https://github.com/dscharrer/innoextract) (license would allow it?!)? I think most have their files from GOG. That would really be the easiest way for the user. Log in to GOG on your smartphone, download the exe and select it the first time you start fheroes2. No data handling on PC.

Just tested it. With the command line variant of innoextract and the EXE from GOG you can get the appropriate files without problems.

I would also like it, by the way, if the ZIP file (or EXE) would also work in the desktop variant and (if no assets available) an openfile dialog opens.

**Comment by oleg-derevenetz** [comment]
@Laserlicht not all users have their HoMM2 installations from GoG, many people have older distributions (for example, those that were still distributed on CD). First we will need to make a more universal version, and only then maybe we will be able to extend it.

**Comment by Laserlicht** [comment]
@oleg-derevenetz Of course. One step after the other.

But I think the Play Store App should keep the entry barrier as low as possible. Many people barely use the classic PC/Laptop these days (hard to believe 😅). For them, installing, copying and zipping on the PC is of course a barrier.

**Comment by oleg-derevenetz** [comment]
@Laserlicht well, at least they definitely have a PC since they bought the game that runs on PC only :) AFAIK there is still no way to run the GoG HoMM2 distribution on Android :)

**Comment by Laserlicht** [comment]
@oleg-derevenetz I was just interested... 😅

Buying on an Android seems to be possible (but I already have it 😉). Download of the EXE also.

<img src="https://user-images.githubusercontent.com/13953785/209247541-fbbade3d-0468-4ec7-ad6d-99966aa0ec9f.jpg" data-canonical-src="https://user-images.githubusercontent.com/13953785/209247541-fbbade3d-0468-4ec7-ad6d-99966aa0ec9f.jpg" width="200" height="400" />
<img src="https://user-images.githubusercontent.com/13953785/209247968-6625bcda-d7e0-48d5-b54f-11c277b4a773.jpg" data-canonical-src="https://user-images.githubusercontent.com/13953785/209247968-6625bcda-d7e0-48d5-b54f-11c277b4a773.jpg" width="200" height="400" />

**Comment by ihhub** [comment]
We cannot release Android version to Google Play Store by 2 reasons:
- missing code to handle files from the original game. At very least we need to add Java code to check that installation path is empty and download a demo version of the game
- missing keyboard support. Using Android keyboard is impossible due to the size of the keyboard overlapping game's UI

These are things which should be done before even starting any integration from GoG.

**Comment by Mr-Bajs** [comment]
Any plans of releasing this on F-droid? Even though initial istallation would require som manual handling with the original files, but it would be more convienient to keeping up to date with the latest fheroes2 release on Android though.

**Comment by oleg-derevenetz** [comment]
> Even though initial istallation would require som manual handling with the original files

Now this should be easier for an end user because of new "toolset" activity.

> Any plans of releasing this on F-droid?

Maybe when we get to that. There are no specific deadlines yet.

**Comment by oleg-derevenetz** [comment]
@ihhub What version of Android Studio you using? Project builds just fine with Electric Eel 2022.1.1 Patch 1.

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I use:
```
Android Studio Dolphin | 2021.3.1 Patch 1
Build #AI-213.7172.25.2113.9123335, built on September 30, 2022
Runtime version: 11.0.13+0-b1751.21-8125866 amd64
```

**Comment by ihhub** [comment]
I am updating Android Studio to see if it helps.

**Comment by oleg-derevenetz** [comment]
@ihhub New Android SDKs want new build tools, which in turn want new Gradle plugin, which in turn want new Android Studio, because older AS are not able to work with newer plugins... :)

**Comment by ihhub** [comment]
But was what the reason to update new tools?

**Comment by oleg-derevenetz** [comment]
> But what was the reason to update new tools?

I updated the target Android SDK to the latest one (API 33). Previously we used the penultimate one (31). It is recommended to always use the latest SDK, and Google Play even enforces this by rejecting apps built with older SDKs (older by 2 years or so).

**Comment by ihhub** [comment]
Thanks for the explanation!

**Comment by oleg-derevenetz** [comment]
@ihhub

https://developer.android.com/google/play/requirements/target-sdk

**Comment by ihhub** [comment]
Hi @oleg-derevenetz , I believe we covered all basic aspect of Android release so we can close this issue. What do you think?

**Comment by oleg-derevenetz** [comment]
@ihhub yes, it seems so :)
